cogs/boolean_utils.py

```diff
@@ -0,0 +1,395 @@
+    # Post-fix version of the original expression
+    post_exp = ""
+    # Set of all the variables
+    var_set = set()
+    # Stack needed to verify correctness of expr
+    var_stack = []
+    # Stack to manage operator precedence
+    op_stack = []
+    # The formatted truth table to be sent on discord
+    result_formatted = ""
+    # Flag and message in case an error occurs
+    error = False
+    error_msg = ""
+
+    # Defines operator precedence
+    prec_dict = {
+        '!' : 0, # NOT
+        '*' : 1, # AND
+        '^' : 2, # XOR
+        '+' : 3, # OR
+        '(' : 8,
+        ')' : 9,
+        -1  : 10 # op_stack is empty, skip
+    }
+
+    def __init__(self, exp):
+        '''Initialize all values'''
+        self.var_set.clear()
+        self.disp_exp = exp
+        self.orig_exp = "".join(exp.split())
+        self.op_stack[:] = []
+        self.var_stack[:] = []
+        self.error = False
+        self.error_msg = ""
+        # Create post-fix expression
+        self.process_exp(self.orig_exp)
+
+    ######################################################################
+    ####### Parse in-fix expression into post-fix ########################
+    ######################################################################
+
+    def process_exp(self, exp):
+        """
+        Processes an in-fix expression and saves a post-fix version
+        of the expression on the class level. The post-fix expression
+        is easier to evaluate in code and makes creating the truth
+        table more efficient.
+        """
+        self.post_exp = ""
+        # Step through chars of booelean expression
+        for char in exp:
+            if self.error == True:
+                return
+            if char == ' ':
+                continue
+            # Alphas are variables
+            elif char.isalpha():
+                self.var_set.add(char)
+                self.var_stack.append(char)
+                self.post_exp += char
+            # Start parens
+            elif char == '(':
+                self.op_stack.append(char)
+            # End parens
+            elif char == ')':
+                # Process all operators until start paren
+                while self.get_top(self.op_stack) != '(':
+                    self.process_an_op()
+                    if self.get_top(self.op_stack) == -1:
+                        self.error = True
+                        self.error_msg = "Unbalanced paren: )"
+                        return
+                # Pop start paren
+                if self.op_stack:
+                    self.op_stack.pop()
+            # If the char is a valid operator then process it
+            elif self.is_valid_op(char):
+                self.process_char(char)
+            else:
+                # we have found an invalid char
+                self.error = True
+                self.error_msg = "{} is not a valid symbol!".format(char)
+        if self.error == True:
+            return
```

Comment:
Would it make more sense to throw an exception here instead?

Reply:
I'm unaware of how the Discord API handles exceptions and I wanted an easy way to display what was wrong with the expression. Throwing an exception would work well, I just don't know how to display that information back to a user in Discord

Reply:
iirc, the discord.py library being used catches all exceptions for you, and there's a way provided by the framework for logging or sending messages on exceptions that I'm not using.

What might be simpler is to wrap the call to this command (in the bot-specific code) with a try catch, and reply back if there was some parsing error with invalid input.